pg_search/src/postgres/types.rs

```diff
@@ -690,13 +690,52 @@ impl TryFrom<TantivyValue> for u64 {
     }
 }
 
+// Classifies NUMERIC: F64 for ±(2^53-2), else I64/U64. Conservative boundary avoids JSON precision loss.
+fn classify_numeric_for_tantivy(numeric: &pgrx::AnyNumeric) -> (bool, bool) {
+    // Use conservative boundary (2^53-2) not (2^53-1) to avoid precision loss during JSON roundtrip.
+    const MAX_SAFE_FOR_JSON: i64 = (1i64 << 53) - 2;
+    const MIN_SAFE_FOR_JSON: i64 = -MAX_SAFE_FOR_JSON;
+
+    // Try to extract as i64 first (covers most common cases)
+    if let Ok(i64_val) = TryInto::<i64>::try_into(numeric.clone()) {
+        // Check if it's within JSON-safe F64 range (excluding boundary values)
+        if i64_val > MIN_SAFE_FOR_JSON && i64_val <= MAX_SAFE_FOR_JSON {
+            return (true, false); // Use F64
+        }
+        return (false, true); // Use I64 to preserve precision
+    }
+
+    // If it doesn't fit in i64, try u64
+    if let Ok(u64_val) = TryInto::<u64>::try_into(numeric.clone()) {
+        if u64_val <= MAX_SAFE_FOR_JSON as u64 {
+            return (true, false); // Use F64
+        }
+        return (false, false); // Use U64 to preserve precision
+    }
+
+    // If it's not an integer, it's already a float - use F64
+    (true, false)
+}
```

Comment:
This method should be inlined, or should directly return an `OwnedValue`, as that would simplify the return type, and for directly executing the relevant casts, rather than executing them once here and then again in the caller.